core/trino-main/src/main/java/io/trino/split/ConnectorAwareSplitSource.java

```diff
@@ -16,29 +16,48 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.util.concurrent.Futures;
 import com.google.common.util.concurrent.ListenableFuture;
+import io.trino.annotation.NotThreadSafe;
 import io.trino.metadata.Split;
 import io.trino.spi.connector.CatalogHandle;
 import io.trino.spi.connector.ConnectorSplit;
 import io.trino.spi.connector.ConnectorSplitSource;
 import io.trino.spi.connector.ConnectorSplitSource.ConnectorSplitBatch;
+import jakarta.annotation.Nullable;
 
 import java.util.List;
 import java.util.Optional;
 
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
 import static io.airlift.concurrent.MoreFutures.toListenableFuture;
 import static java.util.Objects.requireNonNull;
 
+/**
+ * Adapts {@link ConnectorSplitSource} to {@link SplitSource} interface.
+ * <p>
+ * Thread-safety: the implementations is not thread-safe
+ *
+ * @implNote The implementation is internally not thread-safe but also {@link ConnectorSplitSource} is
+ * not required to be thread-safe.
+ */
+@NotThreadSafe
```

Comment:
why not to close at source at `io/trino/execution/scheduler/SourcePartitionedScheduler.java:257` and `io.trino.execution.scheduler.EventDrivenTaskSource.IdempotentSplitSource#getNext`

Reply:
would that apply to FTE scheduler as well?

Reply:
`SourcePartitionedScheduler` is streaming. `EventDrivenTaskSource` is FTE IIRC. @losipiuk knows better